### DISCUSSION TITLE
Java binding: java-tree-sitter

### DISCUSSION THREAD

**Discussion by dabico** [tree-sitter/tree-sitter#3024] [Bindings]
Hello `tree-sitter` community!

I want to share a project I have been working on for the better part of a year. The research group I work for, @seart-group, has a great interest in mining software repositories. One of our ideas was to collect, parse and perform rudimentary analysis on open-source code found across GitHub. In 2022, we released https://github.com/seart-group/DL4SE, a platform that did just that, albeit limited exclusively to Java code. With a desire to expand the mining capabilities of this platform to other languages, we came across `tree-sitter`, and opted to migrate our source code parsing and analysis to use its powerful APIs. There was only one problem: the existing codebase was already written in Java. Since we had no desire to start from scratch in C, and no prior experience working with native code linkage in Java, we opted to turn to the community in hopes of finding existing projects that tackle this particular problem. To our relief, there already was a Java binding: https://github.com/serenadeai/java-tree-sitter. Although this seemed promising at first, we quickly came to the realization that said library lacked many features from the core API that we would need. The fact that it also was not being maintained for a prolonged period did not instil us with any hope. For this reason, we decided to take matters into our own hands and fill the missing gaps from the binding. In 2022, we started development on our fork of the aforementioned binding, localized as a submodule of DL4SE. We worked on it for months, adding the missing features and making enhancements to existing ones. As time went on, we realised that our project had outgrown the original in more ways than one. With a desire to provide the community with a more up-to-date alternative, we published our version of the binding: https://github.com/seart-group/java-tree-sitter. Although a detached fork, we kept the same name, while acknowledging the work of the former development team. At the same time, the code has diverged significantly from the original, especially in the design philosophy and organisation of the source code. While most of the native `tree-sitter` API has been mapped, there are still some open challenges, namely:

- Expanding OS and architecture support past x86_64 Linux/MacOS
- Support for Query predicates like in the JavaScript/Python/Rust bindings

As I write this, version [1.12.0](https://github.com/seart-group/java-tree-sitter/releases/tag/v1.12.0) has just been published. Seeing as though this is my first time working on this type of project, I would greatly appreciate contributions and opinions of any kind. Last but not least, a thank you to all the developers and maintainers for all their amazing work on the `tree-sitter` project and its languages.

**Comment by sogaiu** [comment]
Looks interesting.

I've only done a little with the not-looked-after-that-frequently java-tree-sitter repository but since patching it to work for a few uses I became aware of the following efforts:

* @juli's fork - https://github.com/juli1/java-tree-sitter
* @bonede's effort - https://github.com/bonede/tree-sitter-ng

May be these (and others perhaps) are already known, but FWIW :)
